### PR TITLE
Implement `onnx.Hardmax` lowering

### PR DESCRIPTION
**Pull request by lialan**

### CODE REVIEW

File: lib/Conversion/TorchOnnxToTorch/DefaultDomainGtoP.cpp [hunk 1]
```diff
@@ -1730,4 +1730,57 @@ void mlir::torch::onnx_c::populateDefaultDomainGtoP(
...
+        Value ConstKeepDims = Rewriter.create<Torch::ConstantBoolOp>(
+            Loc, Rewriter.getType<Torch::BoolType>(),
+            Rewriter.getBoolAttr(true));
+        Value Argmax = Rewriter.create<Torch::AtenArgmaxOp>(
+            Loc, ResultType, Input, Axis, ConstKeepDims);
+
+        // scatter_
+        Value OneInt = Rewriter.create<Torch::ConstantIntOp>(
+            Loc, Rewriter.getI64IntegerAttr(1));
+        Rewriter.replaceOpWithNewOp<Torch::AtenScatterValueOp>(
```
Comment (rsuderman) [inline lib/Conversion/TorchOnnxToTorch/DefaultDomainGtoP.cpp:1779]: Its better to use `torch::AtenOneHot` as this directly maps to the operation.

File: lib/Conversion/TorchOnnxToTorch/DefaultDomainGtoP.cpp [hunk 2]
```diff
@@ -1730,4 +1730,57 @@ void mlir::torch::onnx_c::populateDefaultDomainGtoP(
...
                 });
+
+  patterns.onOp(
+      "Hardmax", 13, [](OpBinder Binder, ConversionPatternRewriter &Rewriter) {
+        // def hardmax(tensor, dim=-1):
+        //   zeros = torch.zeros_like(tensor, dtype=torch.int64)
+        //   maximums = torch.argmax(tensor, dim=dim, keepdim=True)
+        //   return zeros.scatter(dim, maximums, value=1)
+
+        Torch::ValueTensorType ResultType;
```
Comment (rsuderman) [inline lib/Conversion/TorchOnnxToTorch/DefaultDomainGtoP.cpp:1741]: Variable names should be `camelCase`. So change `ResultType` to `resultType`. Same for other variables below.

**Review by rsuderman — CHANGES_REQUESTED** [review]

File: lib/Conversion/TorchOnnxToTorch/DefaultDomainGtoP.cpp (hunk 2 above)
Comment (lialan) [reply lib/Conversion/TorchOnnxToTorch/DefaultDomainGtoP.cpp]: updated.

File: lib/Conversion/TorchOnnxToTorch/DefaultDomainGtoP.cpp [hunk 3]
```diff
@@ -1730,4 +1730,57 @@ void mlir::torch::onnx_c::populateDefaultDomainGtoP(
...
+        if (binder.tensorOperand(input) ||
+            binder.s64IntegerAttr(axisValue, "axis") ||
+            binder.tensorResultType(resultType))
+          return failure();
+
+        auto loc = binder.getLoc();
+
+        std::optional<int64_t> axisIntTorch =
+            onnxDtypeIntToTorchDtypeInt(axisValue);
+        if (!axisIntTorch.has_value()) {
```
Comment (rsuderman) [inline lib/Conversion/TorchOnnxToTorch/DefaultDomainGtoP.cpp:1764]: Single line if statements do not require brackets.

**Review by rsuderman — APPROVED** [review]

File: lib/Conversion/TorchOnnxToTorch/DefaultDomainGtoP.cpp (hunk 3 above)
Comment (lialan) [reply lib/Conversion/TorchOnnxToTorch/DefaultDomainGtoP.cpp]: done.

**Pull request merged**